src/uu/sed/src/compiler.rs

```diff
@@ -274,11 +272,150 @@ fn compile_thread(
-// Return the number of addresses encountered.
-fn compile_addresses(_line: &mut ScriptCharProvider, _cmd: &mut Command) -> usize {
-    // TODO: implement address parsing
-    0
+/// Return true if c is a valid character for specifying a context address
+fn is_address_char(c: char) -> bool {
+    matches!(c, '0'..='9' | '/' | '\\' | '$')
+}
+
+/// Compile a command's optional address range into cmd.
+/// Return the number of addresses encountered.
+fn compile_address_range(
+    lines: &ScriptLineProvider,
+    line: &mut ScriptCharProvider,
+    cmd: &mut Command,
+) -> UResult<usize> {
+    let mut n_addr = 0;
+
+    line.eat_spaces();
+    if !line.eol() && is_address_char(line.current()) {
+        if let Ok(addr1) = compile_address(lines, line) {
+            cmd.addr1 = Some(addr1);
+            n_addr += 1;
+        }
+    }
+
+    line.eat_spaces();
+    if n_addr == 1 && !line.eol() && line.current() == ',' {
+        line.advance();
+        line.eat_spaces();
+        if !line.eol() {
+            if let Ok(addr2) = compile_address(lines, line) {
+                cmd.addr2 = Some(addr2);
+                n_addr += 1;
+            }
+        }
+    }
+
+    Ok(n_addr)
+}
+
+/// Compile and return a single range address specification.
+fn compile_address(lines: &ScriptLineProvider, line: &mut ScriptCharProvider) -> UResult<Address> {
+    let mut icase = false;
+
+    if line.eol() {
+        return compilation_error(lines, line, "expected context address");
+    }
+
+    match line.current() {
+        '\\' | '/' => {
+            // Regular expression
+            if line.current() == '\\' {
+                // The next character is an arbitrary delimiter
+                line.advance();
+            }
+            let re = parse_regex(lines, line)?;
+            // Skip over delimiter
+            line.advance();
+
+            line.eat_spaces();
+            if !line.eol() && line.current() == 'I' {
+                icase = true;
+                line.advance();
+            }
+
+            Ok(Address {
+                atype: AddressType::Re,
+                value: AddressValue::Regex(compile_regex(lines, line, &re, icase)?),
+            })
+        }
+        '$' => {
+            line.advance();
+            Ok(Address {
+                atype: AddressType::Last,
+                value: AddressValue::LineNumber(0),
+            })
+        }
+        '+' => {
+            line.advance();
+            let number = parse_number(lines, line)?;
+            Ok(Address {
+                atype: AddressType::RelLine,
+                value: AddressValue::LineNumber(number),
+            })
+        }
+        c if c.is_ascii_digit() => {
+            let number = parse_number(lines, line)?;
+            Ok(Address {
+                atype: AddressType::Line,
+                value: AddressValue::LineNumber(number),
+            })
+        }
+        _ => panic!("invalid context address"),
```

Comment:
panic a bit violent in rust :)
also, you might want to add the line info to help with debugging

Reply:
It's an internal error; it shouldn't happen. The code reaches this point only when `is_address_char` has detected one of the preceding characters.